packages/frontend/src/features/sqlRunner/components/SqlEditor.tsx

```diff
@@ -237,17 +259,73 @@ export const SqlEditor: FC<{
         [language, quoteChar, tablesData],
     );
 
-    const onMount: OnMount = useCallback((editor, monaco) => {
-        editorRef.current = editor;
-        editor.addCommand(monaco.KeyMod.CtrlCmd | monaco.KeyCode.Enter, () => {
-            // When the editor is mounted, the onSubmit callback should be set to the latest value, otherwise it will be set to the initial value on the first render
-            onSubmitRef.current?.();
-        });
-    }, []);
+    const monaco = useMonaco();
+    const decorationsCollectionRef =
+        useRef<editor.IEditorDecorationsCollection | null>(null); // Ref to store the decorations collection
 
-    const onChange: OnChange = debounce((value: string | undefined) => {
-        dispatch(setSql(value ?? ''));
-    }, DEBOUNCE_TIME);
+    const onMount: OnMount = useCallback(
+        (editorObj, monacoObj) => {
+            editorRef.current = editorObj;
+            decorationsCollectionRef.current =
+                editorObj.createDecorationsCollection(); // Initialize the decorations collection
+            editorObj.addCommand(
+                monacoObj.KeyMod.CtrlCmd | monacoObj.KeyCode.Enter,
+                () => {
+                    onSubmit?.();
+                },
+            );
+        },
+        [onSubmit],
+    );
+
+    useEffect(() => {
+        // remove any existing decorations
+        if (decorationsCollectionRef.current) {
+            decorationsCollectionRef.current.set([]);
+        }
+        if (!editorRef.current || !monaco || !decorationsCollectionRef.current)
+            return;
+        // do nothing if no highlightText is provided
+        if (!highlightText) return;
+        // if no end, highlight only the start + 1 character
+        if (!highlightText.end) {
+            highlightText.end = {
+                line: highlightText.start.line,
+                char: highlightText.start.char + 1,
+            };
+        }
+        const range = new monaco.Range(
+            highlightText.start.line,
+            highlightText.start.char,
+            highlightText.end.line,
+            highlightText.end.char,
+        );
+        const newDecorations = [
+            {
+                range,
+                options: {
+                    inlineClassName: 'editorError',
+                },
+            },
+        ];
+
+        // Update decorations using the decorations collection
+        decorationsCollectionRef.current.set(newDecorations);
+    }, [sql, monaco, highlightText]);
+
+    useEffect(() => {
```

Comment:
We can avoid using the `useEffect` and the internal state `value`. 

I recommend doing this instead (since we have RTK for the SQL state): 

```tsx
  onChange={onChange}
```

```tsx
 const onChange: OnChange = debounce((value: string | undefined) => {
        dispatch(setSql(value ?? ''));
        if (resetHighlightError && highlightText) {
            resetHighlightError();
        }
    }, DEBOUNCE_TIME);
```

As you start typing (and when there's an highlight text error) it will reset, removing then the need for this `useEffect`. It also has improved performance for the SQL typing

Reply:
The issue with this is that the reset highlightError is debounced and will have a delay clearing, so as a user types (and fixing the error) it will remain highlighted until the debounce fires. The current useEffect will clear the highlight right away while still debouncing the dispatch.

Reply:
oops! 

```tsx

    const debouncedSetSql = useMemo(
        () =>
            debounce((value: string) => dispatch(setSql(value)), DEBOUNCE_TIME),
        [dispatch],
    );

    const onChange: OnChange = useCallback(
        (value: string | undefined) => {
            if (highlightText && resetHighlightError) {
                resetHighlightError();
            }
            debouncedSetSql(value ?? '');
        },
        [debouncedSetSql, highlightText, resetHighlightError],
    );

```

This way we avoid use effect pitfalls when not providing all deps referenced in it. 

What do you think?